### ISSUE TITLE
Custom policy that only samples from legal actions - PPO2 [question]

### ISSUE THREAD

**Issue by maxvfischer** [hill-a/stable-baselines#212]
Hi,
I'm currently trying to implement PPO2. My action space is discrete (144), but only some of the actions are legal in a given state. The legal actions varies depending on the state, so for example in one state actions [1, 15, 133] is legal to do and in another state [44, 53, 99, 102] is legal.

I have access to the legal actions, so I can pass a multi-hot-vector (legal = 1, illegal = 0) to the model that can be used to filter out some actions.

My question is where and how I should do it, do you have any recommendations?

Here is my current custom policy, without removing the illegal moves:
```
class CustomPolicy(ActorCriticPolicy):
    def __init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=False, **kwargs):
        super(CustomPolicy, self).__init__(sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=reuse, scale=True)

        with tf.variable_scope("model", reuse=reuse):
            extracted_features = self.ppo_cnn(self.processed_obs, **kwargs)
            extracted_features = tf.layers.flatten(extracted_features)

            activ = tf.nn.elu
            pi_h = extracted_features
            for i, layer_size in enumerate([128, 128, 128]):
                pi_h = activ(tf.layers.dense(pi_h, layer_size, name='pi_fc' + str(i)))
            pi_latent = pi_h

            vf_h = extracted_features
            for i, layer_size in enumerate([32, 32]):
                vf_h = activ(tf.layers.dense(vf_h, layer_size, name='vf_fc' + str(i)))
            value_fn = tf.layers.dense(vf_h, 1, name='vf')
            vf_latent = vf_h

            self.proba_distribution, self.policy, self.q_value = \
                self.pdtype.proba_distribution_from_latent(pi_latent, vf_latent, init_scale=0.01)

        self.value_fn = value_fn
        self.initial_state = None
        self._setup_init()

    def step(self, obs, state=None, mask=None, legal_actions=None):
        action, value, neglogp = self.sess.run([self.action, self._value, self.neglogp],
                                                       {self.obs_ph: obs, self.legal_actions: legal_actions})

        return action, value, self.initial_state, neglogp

    def proba_step(self, obs, state=None, mask=None):
        return self.sess.run(self.policy_proba, {self.obs_ph: obs})

    def value(self, obs, state=None, mask=None):
        return self.sess.run(self._value, {self.obs_ph: obs})

    def ppo_cnn(self, board_state, **kwargs):
        activ = tf.nn.relu
        layer_1 = tf.layers.conv2d(inputs=board_state, filters=128, kernel_size=2, padding="same", activation=tf.nn.relu)
        layer_2 = tf.layers.conv2d(inputs=layer_1, filters=128, kernel_size=3, padding="same", activation=tf.nn.relu)
        layer_3 = tf.layers.conv2d(inputs=layer_2, filters=128, kernel_size=3, padding="same", activation=tf.nn.relu)
        layer_4 = conv_to_fc(layer_3)

        return layer_4

```

**Comment by araffin** [comment]
Similar to #108

**Comment by H-Park** [comment]
Look into TensorFlow's stop gradient method. This does exactly what you're asking.

**Comment by araffin** [comment]
I will close this issue in favor of #351
